src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs

```diff
@@ -131,10 +131,72 @@ public PackageArchiveReader BuildPackage(PackageBuilder builder, string outputPa
             return new PackageArchiveReader(outputPath);
         }
 
+        /// <summary>
+        /// Writes the resolved NuSpec file to the package output directory.
+        /// </summary>
+        /// <param name="builder">The package builder</param>
+        private void WriteResolvedNuSpecToPackageOutputDirectory(PackageBuilder builder)
+        {
+            var outputPath = GetOutputPath(builder, _packArgs, false, builder.Version);
+            Directory.CreateDirectory(Path.GetDirectoryName(outputPath));
+
+            var resolvedNuSpecOutputPath = Path.Combine(
+                Path.GetDirectoryName(outputPath), 
+                new VersionFolderPathResolver(outputPath).GetManifestFileName(builder.Id, builder.Version));
+
+            _packArgs.Logger.LogMinimal(string.Format(CultureInfo.CurrentCulture, Strings.Log_PackageCommandInstallPackageToOutputPath, "NuSpec", resolvedNuSpecOutputPath));
+
+            if (string.Equals(_packArgs.Path, resolvedNuSpecOutputPath, StringComparison.OrdinalIgnoreCase))
+            {
+                throw new PackagingException(NuGetLogCode.NU5001, string.Format(CultureInfo.CurrentCulture, Strings.Error_WriteResolvedNuSpecOverwriteOriginal, _packArgs.Path));
+            }
+
+            var manifest = new Manifest(new ManifestMetadata(builder), null);
+            using (Stream stream = new FileStream(resolvedNuSpecOutputPath, FileMode.Create))
+            {
+                manifest.Save(stream);
+            }
+        }
+
+        /// <summary>
+        /// Writes the sha512 package hash file to the package output directory
+        /// </summary>
+        /// <param name="builder">The package builder</param>
+        private void WriteSHA512PackageHash(PackageBuilder builder)
+        {
+            var outputPath = GetOutputPath(builder, _packArgs, false, builder.Version);
+            Directory.CreateDirectory(Path.GetDirectoryName(outputPath));
+
+            var sha512OutputPath = Path.Combine(outputPath + ".sha512");
+            var sha512TempOutputPath = Path.Combine(NuGetEnvironment.GetFolderPath(NuGetFolderPath.Temp), Path.GetFileName(sha512OutputPath));
+
+            _packArgs.Logger.LogMinimal(string.Format(CultureInfo.CurrentCulture, Strings.Log_PackageCommandInstallPackageToOutputPath, "SHA512", sha512OutputPath));
+
+            byte[] sha512hash;
+            var cryptoHashProvider = new CryptoHashProvider("SHA512");
+            using (var fileStream = new FileStream(outputPath, FileMode.Open, FileAccess.Read))
+            {
+                sha512hash = cryptoHashProvider.CalculateHash(fileStream);
+            }
+
+            FileUtility.Delete(sha512TempOutputPath);
+            File.WriteAllText(sha512TempOutputPath, Convert.ToBase64String(sha512hash));
+
+            FileUtility.Delete(sha512OutputPath);
+            FileUtility.Move(sha512TempOutputPath, sha512OutputPath);
```

Comment:
Use FileUtility.Replace, it creates the temp file and does all of this for you.

https://github.com/NuGet/NuGet.Client/blob/ceb01f8696f29c8fc2c871180c1e552bd98ac154/src/NuGet.Core/NuGet.Common/PathUtil/FileUtility.cs#L83